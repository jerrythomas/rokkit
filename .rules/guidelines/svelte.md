# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
