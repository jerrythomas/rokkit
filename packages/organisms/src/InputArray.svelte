<script>
	import List from './List.svelte'
	import FormFields from './FormFields.svelte'

	let className = ''
	export { className as class }
	export let path = []
	export let value = []
	export let schema = []
	export let using = {}
	export let actions = {}
	export let navigator = List
	let item

	function updatePath(item) {
		let index = value.indexOf(item)

		return index
			? [
					...path.slice(0, path.length - 1),
					path.slice(path.length - 1) + '[' + index + ']'
			  ]
			: []
	}
	$: location = updatePath(item)
</script>

<section class="flex {className}">
	{#if navigator}
		<svelte:component this={navigator} items={value} bind:value={item} />
	{/if}

	<wrap-items>
		<!-- Array actions add, remove -->
		<FormFields bind:value={item} {schema} {using} path={location} />
	</wrap-items>
</section>
