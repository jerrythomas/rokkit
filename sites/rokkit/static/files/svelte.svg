<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
