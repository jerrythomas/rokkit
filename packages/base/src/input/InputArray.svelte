<script>
	import { createEventDispatcher } from 'svelte'
	import Tabs from './Tabs.svelte'
	import FormFields from './FormFields.svelte'

	const dispatch = createEventDispatcher()

	export let id
	export let value
	export let fields
	export let allowAdd = false
	export let allowClose = false
	// export let limit
	export let label

	function onChange() {
		value = [...value]
		dispatch('change', { value })
	}

	let item = value[0]
</script>

{#if Array.isArray(value)}
	<tab-edit {id} class="flex flex-col w-full">
		{#if label}
			<label for={id}>{label}</label>
		{/if}
		<Tabs bind:items={value} {allowAdd} {allowClose} bind:activeItem={item} />
		<content>
			<FormFields bind:data={item} {fields} on:change={onChange} />
		</content>
	</tab-edit>
{:else}
	<error>Property `value` should be an array of objects.</error>
{/if}
