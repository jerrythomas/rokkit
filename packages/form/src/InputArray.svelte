<script>
	import { createEventDispatcher } from 'svelte'
	import FormFields from './FormFields.svelte'
	import {Tabs} from '@jerrythomas/spice-list'

	const dispatch = createEventDispatcher()

	export let value
	export let fields
	export let allowAdd = false
	export let allowClose = false

	export let label

	function onChange() {
		value = [...value]
		dispatch('change', { value })
	}

	let item = value[0]
</script>

{#if Array.isArray(value)}
	<Tabs
		bind:items={value}
		bind:activeItem={item}
		title={label}
		{allowAdd}
		{allowClose}
	>
		<FormFields bind:data={item} {fields} on:change={onChange} />
	</Tabs>
{:else}
	<error>Property `value` should be an array of objects.</error>
{/if}
