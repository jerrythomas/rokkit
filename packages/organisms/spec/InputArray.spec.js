import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'

import InputArray from '../src/InputArray.svelte'

describe('InputArray.svelte', () => {
	beforeEach(() => cleanup())

	it('Should render using default field mapping', () => {
		const { container } = render(InputArray)
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('Should render using field mappings', () => {})
	it('Should render items using custom component', () => {})
	it('Should expand and collapse', () => {})
	it('Should autoclose others', () => {})
	it('Should pass select and change events', () => {})
})
