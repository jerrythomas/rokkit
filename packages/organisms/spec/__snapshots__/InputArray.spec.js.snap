// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputArray.svelte > should render using default field mapping 1`] = `
<body>
  <div>
    <section
      class="flex "
    >
      <list
        class="flex flex-col w-full flex-shrink-0 select-none list"
        role="listbox"
        tabindex="0"
      >
         
      </list>
      
       
      <wrap-items>
        <error>
          Invalid schema. Expected schema to include an 'elements' array.
        </error>
        
      </wrap-items>
    </section>
  </div>
</body>
`;
